### ISSUE TITLE
Rule proposal: avoid unhandled rejections

### ISSUE THREAD

**Issue by MarcCelani-at** [typescript-eslint/typescript-eslint#6418]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

There is no rule with the explicit goal of detecting and preventing unhandled rejections. no-floating-promises does prevent one class of unhandled rejections (floating promises can cause unhandled rejections), but there are other cases that are not covered and should not be covered (see discussion in #6416 and fail cases below)

### Fail Cases

```typescript
// This is covered by no-floating-promises, but should be an error for this rule as well
async function failCaseFloatingPromise() {
    Promise.reject();
}

async function failCasePromiseAll() {
    const a = Promise.reject();
    const b = Promise.resolve(42);
    await Promise.all([a, b]);
}

async function failCasePromiseRace() {
    const a = Promise.reject();
    const b = Promise.resolve(42);
    await Promise.race([a, b]);
}

async function failCaseAsyncGap() {
    const a = Promise.reject();
    const b = Promise.resolve(42);
    await b;
    await a;
}

// This is the motivating example that spurred this feature request.
async function failSimplifiedRealWorldScenario() {
    const primaryQueryPromise = queryPrimaryAsync();
    const secondaryQueryPromise = querySecondaryAsync();
    try {
        primaryQueryResult = await primaryQueryPromise;
        secondaryQueryPromise.then(
            (secondaryQueryResult) => {
                compareResultsAndLog(primaryQueryResult, secondaryQueryResult);
            },
            (_) => { },
        );
    } catch {
    }
}
```

### Pass Cases

```typescript
async function passCasePromiseAllSettled() {
    const a = Promise.reject();
    const b = Promise.resolve(42);
    await Promise.allSettled([a, b]);
}

async function passCaseManyPromisesSequentiallyAwaitedWithNoAsyncGap() {
    await Promise.resolve(42);
    await Promise.reject();
}

async function passCaseThenOrCatchCallback() {
    // Similar rules as no-floating-promises, setting up catch callbacks or
    // then callbacks with rejection handlers before first await is safe.
    Promise.reject().catch((_) => {});
    Promise.reject().then((value) => { }, (error) => { });
    await Promise.resolve(42);
}

// Prevents unhandled rejections in fire and forget methods.
// To use: `void fireAndForgetHelperMethod(() => { /* your code */ });`
async function fireAndForgetHelperMethod(fn) {
  return fn().catch((_) => {});
}

// Similar rules as no-floating-promises
async function passCaseExplicitVoidForFireAndForget() {
  void fireAndForgetHelperMethod(async function() { });
}

// Ideally, the type system would be able to detect when a promise cannot be rejected (typed here
// as an "UnrejectablePromise"), and developers could use the `as` keyword to typecast when the
// type checker cannot infer this itself. Then, this rule would not apply to UnrejectablePromises.
async function getUnrejectablePromise(): UnrejectablePromise<number> {
    return await Promise.resolve(42);
}

async function passCaseUnrejectablePromise() {
    const unrejectablePromise = getUnrejectablePromise();
}
```

### Additional Info

failSimplifiedRealWorldScenario is based off a real world bug that caused crashes in our nodejs service. This code is buggy for two reasons.

1. If queryPrimaryAsync throws, no callback is set on secondaryQueryPromise and there will be an unhandled rejection if querySecondaryAsync throws.
2. If querySecondaryAsync rejects before queryPrimaryAsync resolves, there will be an unhandled rejection because the callback has not been set in time.

Fixing this bug requires setting a catch callback on secondaryQueryPromise before awaiting the primaryQueryPromise. But in practice, developers like to use utility functions for doing things like chaining callbacks to swallow exceptions. Hence the desire for something like an UnrejectablePromise.

**Comment by MarcCelani-at** [comment]
Moving discussion from #6416 , @bradzacher said:

>So if you define a helper function like preventUnhandledRejections, then a rule will continue to report on those promises. If you reassign a promise or pass it through a function then a rule can't detect the change and will report on those cases.
Which makes the rule dead in the water in a general purpose plugin like ours - we'd just create noise.

This does feel like the type of rule that would be opt in. But I disagree that this would just create noise. At least in our fairly modern large scale monorepository, we heavily use await and most async code is written sequentially or with a fireAndForget utility that could play very nicely with a void hint like in no-floating-promises (added in an example). In our monorepo, there are...

20k instances of `await`
750 instances of a utility to fireAndForget, which could probably play nice with this rule with a `void` keyword.
500 instances of `Promise.all`, many of which should probably migrate to `Promise.allSettled`
35 instances of `Promise.allSettled`
30 instances of `Promise.race`, many of which we would probably inspect and ignore the error for

**Comment by bradzacher** [comment]
Thanks for the in-depth proposal!

The big thing we look for when accepting new rules in this project is "do we think this rule is something the community will broadly adopt or find value in".
We're at the scale now where we have a large number of rules to maintain with our small volunteer team, and we cover a large chunk of issues that users face, so we provide a solid offering.

First up let me being by saying that this is the first time that I've personally seen anyone actively worried about out-of-order rejection. It's an interesting thing to think about, and is definitely a valid problem in code that uses `async`/`await` with multiple promises and "backgrounds" one promise to `await` another one.

However from my experience this sort of usecase is really niche! From what I've seen most everyday async code deals with just one promise and/or deals with promises that very unlikely to reject (or will only reject in "FUBAR" cases) - the former is not impacted by this and the latter generally doesn't need to consider it.

In terms of `Promise.all` and `Promise.race` - I lead you astray in the previous thread. In a nutshell those utilities attach catch handlers to each promise they're passed - which nullifies unhandled promise rejection errors. It doesn't mean that slow rejections are "handled" - just that they won't leak unhandled rejection errors.

Example code to show this to be true:
```ts
async function foo() {
  try {
    const res = await Promise.race([
      Promise.reject(new Error('instant error')),
      new Promise(resolve => setTimeout(resolve, 1000)),
      new Promise((_, reject) => setTimeout(() => reject(new Error('delayed error')), 2000)),
    ]);
    console.log(res);
  } catch (e) {
    console.log(e);
  }
}

foo();
```

Which narrows the scope of this rule significantly to just "code that `await`s multiple promises where one is 'backgrounded'". As mentioned this is just super niche - so much so that I don't think it makes sense as a rule within our general-purpose plugin. Not saying it's a bad rule at all, just that in the general case, which we try to address, the rule would not be useful.

I'd love to hear what other maintainers think, but at least from me I don't think this rule belongs within this project.
cc @JoshuaKGoldberg / @Josh-Cena

**Comment by JoshuaKGoldberg** [comment]
+1 to bradzacher's post. Even if this kind of problem is something universally applicable enough to warrant inclusion typescript-eslint -which I don't see enough evidence for- it'd be a very difficult rule to get right. I think the best step forward would be to make a community plugin with a [custom typed rule](https://typescript-eslint.io/custom-rules) and see how folks use it there. That'd give you the benefit of being able to make quick iterations & breaking changes without getting bogged down by the typescript-eslint release cycle!

**Comment by Josh-Cena** [comment]
I don't have strong opposition to the rule, but I'm interested in what the rule checks against specifically. That any promise created must not be followed by the awaiting of an unrelated promise? That requires us to build a dependency graph, e.g. to tell that `Promise.allSettled([a, b])` depends on the promises `a` and `b`.

**Comment by bradzacher** [comment]
@Josh-Cena - the problem is best represented like this:
```ts
function backgroundWork() { return Promise.reject('oops') }
function foregroundWork() { return new Promise(resolve => setTimeout(resolve, 1000)) }

async function foo() {
  const backgroundedPromise = backgroundWork();
  const foregroundResult = await foregroundWork();

  try {
    await backgroundedPromise
  } catch (e) {
    console.log('caught', e);
  }
}

foo();
```

If you run this code you'll get an error log because the rejection of `backgroundPromise` occurred and wasn't handled. The reason for the log - because engines use a heuristic to detect if a promise rejection is handled - essentially "If there is no immediate `catch` handler, Wait until the end of the current event loop - if there is no handler at that point, error".

This isn't a **huge** problem because after all it's just a log line.
However consider the case where `foregroundWork` rejects instead of resolving - now there will never be any error handler for `backgroundedPromise` at all! Its error will truly be unhandled because the try/catch will never execute.

So the check would be more of a control-flow check - i.e. "ensure that `backgroundedPromise` has a `catch` handler if there are branches in the current function that might terminate before `await`ing it".

But the problem with such a rule is that there's no way to statically tell for certain whether or not `foregroundWork()` will ever actually throw (halting problem!) because there's no way in TS to represent a promise with a catch handler. So even if `foregroundWork` is like
```ts
async function foregroundWork() {
  try {
    const prom = new Promise(...);
    await prom;
  } catch {}
}
```
A rule would have to assume that the function might throw, and treat it as a terminating branch.

**Comment by bradzacher** [comment]
Thinking more - another problem with things is that there is no way to truly enforce that a promise will never reject.

For example:
```ts
function handleError() {
  throw new Error('woopsie');
}

const prom = foregroundWork().catch(handleError);
```

This code statically looks 100% fine - but without analysing the code more deeply you can't see that the promise still actually rejects! Woopsie!

Unfortunately there's just no way to build this rule without making a lot of shaky assumptions - which doesn't suit a general-purpose plugin like ours.

I think we're all in agreement that this rule could work well for some codebases, but it does not meet the bar for a rule we want to accept into this project.
Thanks for filing!
